### ISSUE TITLE
[Feature] Would it be possible to implement an RBF NN in MLX

### ISSUE THREAD

**Issue by sck-at-ucy** [ml-explore/mlx#509]
Could the MLX team consider implementing a native Radial Basis Function (RBF) layer in the MLX framework? An RBF layer would greatly enhance MLX's capabilities in function approximation and interpolation, particularly for applications requiring smooth, localized transformations of inputs. This addition would be beneficial for users seeking to integrate complex, non-linear mappings into their MLX-based machine learning models. This would be particularly useful in scientific computing for scenarios where the loss needs to be  based on data generated with external libraries that cannot be  used directly because autograd would not be able to compute the gradients.

**Comment by angeloskath** [comment]
Are you looking for something like the following or something more complicated?

```python
import mlx.core as mx
import mlx.nn as nn


class RBFLayer(nn.Module):
    def __init__(self, input_dims, hidden_dims, output_dims, bias=True):
        super().__init__()

        # The initialization of the centers and betas depends on the input
        # distribution, but the following should be kind of fine for a standard
        # normal. Also the betas are going to be learnable this way, maybe it
        # is better to freeze them or make sure they are positive by
        # parameterizing them as logbetas.
        self.centers = mx.random.normal((hidden_dims, input_dims))
        self.betas = mx.full((hidden_dims,), 1/input_dims)
        self.projection = nn.Linear(hidden_dims, output_dims, bias=bias)

    def __call__(self, x):
        # Normalize the shape a bit
        shape = x.shape
        x = x.reshape(-1, shape[-1])

        # Pairwise distance
        x2 = 0.5 * x.square().sum(-1)
        c2 = 0.5 * self.centers.square().sum(-1)
        xc = x @ self.centers.T
        dist = x2[:, None] + c2[None, :] - xc

        # Compute the rbf
        # Here we could use mx.softmax(-dist, -1) for normalized rbf
        # or have fixed β_i etc.
        rbf = mx.exp(-(self.betas * dist))

        # Finally gather the outputs
        return self.projection(rbf).reshape(*shape[:-1], -1)

```

**Comment by sck-at-ucy** [comment]
Hi Angele, 

Thank you, this is a great. I will have to experiment with it and fine tune it a bit but it gives me a good starting point and I will reach out for your wisdom if I run into trouble. If I manage to get it working with a couple of examples I am trying right now, I will share it back including any adjustments.

**Comment by angeloskath** [comment]
I am closing the issue but feel free to reopen if something comes up.
